Update to support Python 3.9-3.12

TaperChipmunk32 opened the pull request:
-Updated machine.py to support Python 3.9-3.12
-Updated Ubuntu version to noble (24.04), along with CUDA version to match Ubuntu
-Updated dev docker container to Python 3.9 and Ubuntu noble
-Updated non-dev docker container to Python 3.12 and Ubuntu noble
-Updated CI to build and test Python 3.9-3.12
-Updated transformers support to versions ">=4.38.0,<4.46"

<!-- Reviewable:start -->
- - -
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/sillsdev/machine.py/131)
<!-- Reviewable:end -->

johnml1135 commented:
*[`pyproject.toml` line 67 at r1](https://reviewable.io/reviews/sillsdev/machine.py/131#-O9WPe_oAqte-xkCwsaI:-O9WPe_oAqte-xkCwsaJ:b-fhbh47) ([raw file](https://github.com/sillsdev/machine.py/blob/1c47de3b2062f500535f4acd2448915a9a36a6f8/pyproject.toml#L67)):*
> ```TOML
> transformers = "^4.34.0, <4.38.0"
> datasets = "^2.4.0"
> sacremoses = "^0.0.53"
> ```

Shouldn't these other packages be updated as well?  Now is the time to see if we can update all the packages.


<!-- Sent from Reviewable.io -->

mshannon-sil reviewed (commented):
> Reviewed 16 of 16 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/sillsdev/machine.py/131)* status: all files reviewed, 4 unresolved discussions (waiting on @ddaspit, @johnml1135, and @TaperChipmunk32)

___
*[`.devcontainer/dockerfile` line 38 at r1](https://reviewable.io/reviews/sillsdev/machine.py/131#-O9kPlss15Dog9dlJHqz:-O9kPlstDBa2m-IicqaA:bp9gdwb) ([raw file](https://github.com/sillsdev/machine.py/blob/1c47de3b2062f500535f4acd2448915a9a36a6f8/.devcontainer/dockerfile#L38)):*
> ```
>     && pip install poetry==${POETRY_VERSION} black pipenv virtualenv clearml \
>     && poetry config virtualenvs.in-project true \
>     && pip install cffi
> ```

What package is depending on cffi? And is there a reason it needs to be in the dockerfile rather than pyproject.toml?
___
*[`.devcontainer/dockerfile-Python312` line 29 at r1](https://reviewable.io/reviews/sillsdev/machine.py/131#-O9kT65y6BMdZ_JPgMQK:-O9kT65y6BMdZ_JPgMQL:bgtajxz) ([raw file](https://github.com/sillsdev/machine.py/blob/1c47de3b2062f500535f4acd2448915a9a36a6f8/.devcontainer/dockerfile-Python312#L29)):*
> ```
> 
> # make some useful symlinks that are expected to exist
> RUN ln -sfn /usr/bin/python${PYTHON_VERSION} /usr/bin/python3  & \
> ```

We're keeping the 3.9 version as the default, so you can probably just remove the 3.12 version of the dev docker container, unless someone else has a reason we should keep it.
___
*[`.vscode/PythonImportHelper-v2-Completion.json` line 0 at r1](https://reviewable.io/reviews/sillsdev/machine.py/131#-O9kTNn0-PcUMpMlds-q:-O9kTNn0-PcUMpMlds-r:bltw82y) ([raw file](https://github.com/sillsdev/machine.py/blob/1c47de3b2062f500535f4acd2448915a9a36a6f8/.vscode/PythonImportHelper-v2-Completion.json#L0)):*
This file probably shouldn't be needed.


<!-- Sent from Reviewable.io -->

TaperChipmunk32 commented:
> > Reviewed 16 of 16 files at r1, all commit messages.
> > _[Reviewable](https://reviewable.io/reviews/sillsdev/machine.py/131)_ status: all files reviewed, 4 unresolved discussions (waiting on @ddaspit, @johnml1135, and @TaperChipmunk32)
> 
> _[`.devcontainer/dockerfile` line 38 at r1](https://reviewable.io/reviews/sillsdev/machine.py/131#-O9kPlss15Dog9dlJHqz:-O9kPlstDBa2m-IicqaA:bp9gdwb) ([raw file](https://github.com/sillsdev/machine.py/blob/1c47de3b2062f500535f4acd2448915a9a36a6f8/.devcontainer/dockerfile#L38)):_
> 
> > ```
> >     && pip install poetry==${POETRY_VERSION} black pipenv virtualenv clearml \
> >     && poetry config virtualenvs.in-project true \
> >     && pip install cffi
> > ```
> 
> What package is depending on cffi? And is there a reason it needs to be in the dockerfile rather than pyproject.toml?
> 
> _[`.devcontainer/dockerfile-Python312` line 29 at r1](https://reviewable.io/reviews/sillsdev/machine.py/131#-O9kT65y6BMdZ_JPgMQK:-O9kT65y6BMdZ_JPgMQL:bgtajxz) ([raw file](https://github.com/sillsdev/machine.py/blob/1c47de3b2062f500535f4acd2448915a9a36a6f8/.devcontainer/dockerfile-Python312#L29)):_
> 
> > ```
> > 
> > # make some useful symlinks that are expected to exist
> > RUN ln -sfn /usr/bin/python${PYTHON_VERSION} /usr/bin/python3  & \
> > ```
> 
> We're keeping the 3.9 version as the default, so you can probably just remove the 3.12 version of the dev docker container, unless someone else has a reason we should keep it.
> 
> _[`.vscode/PythonImportHelper-v2-Completion.json` line 0 at r1](https://reviewable.io/reviews/sillsdev/machine.py/131#-O9kTNn0-PcUMpMlds-q:-O9kTNn0-PcUMpMlds-r:bltw82y) ([raw file](https://github.com/sillsdev/machine.py/blob/1c47de3b2062f500535f4acd2448915a9a36a6f8/.vscode/PythonImportHelper-v2-Completion.json#L0)):_ This file probably shouldn't be needed.

I'll check if I can move cffi to pyproject.toml, but I believe there was a reason I put it in the container instead. I'll look into it again to be sure.

Also, I did not realize the import helper got pushed, I'll add it to .gitignore.

TaperChipmunk32 commented:
> _[`pyproject.toml` line 67 at r1](https://reviewable.io/reviews/sillsdev/machine.py/131#-O9WPe_oAqte-xkCwsaI:-O9WPe_oAqte-xkCwsaJ:b-fhbh47) ([raw file](https://github.com/sillsdev/machine.py/blob/1c47de3b2062f500535f4acd2448915a9a36a6f8/pyproject.toml#L67)):_
> 
> > ```toml
> > transformers = "^4.34.0, <4.38.0"
> > datasets = "^2.4.0"
> > sacremoses = "^0.0.53"
> > ```
> 
> Shouldn't these other packages be updated as well? Now is the time to see if we can update all the packages.

I was going to keep transformers downgraded because of some changes made to the NLLBTokenizers in later versions, but I will work on upgrading it instead. 

I've also started looking into upgrading the other packages mentioned.

TaperChipmunk32 commented:
> ## [Codecov](https://app.codecov.io/gh/sillsdev/machine.py/pull/131?dropdown=coverage&src=pr&el=h1&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=sillsdev) Report
> 
> Attention: Patch coverage is `84.31373%` with `8 lines` in your changes missing coverage. Please review.
> 
> > Project coverage is 88.12%. Comparing base [(`e0e3f08`)](https://app.codecov.io/gh/sillsdev/machine.py/commit/e0e3f081d57f5a998949bd6cc60f345fa0f74789?dropdown=coverage&el=desc&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=sillsdev) to head [(`186527d`)](https://app.codecov.io/gh/sillsdev/machine.py/commit/186527dc4cf5a79db448eddfc882ec031bfb4c89?dropdown=coverage&el=desc&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=sillsdev).
> > Report is 7 commits behind head on main.
> 
> [Files with missing lines](https://app.codecov.io/gh/sillsdev/machine.py/pull/131?dropdown=coverage&src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=sillsdev) 	Patch % 	Lines
> [machine/translation/thot/thot_smt_model_trainer.py](https://app.codecov.io/gh/sillsdev/machine.py/pull/131?src=pr&el=tree&filepath=machine%2Ftranslation%2Fthot%2Fthot_smt_model_trainer.py&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=sillsdev#diff-bWFjaGluZS90cmFuc2xhdGlvbi90aG90L3Rob3Rfc210X21vZGVsX3RyYWluZXIucHk=) 	71.42% 	[2 Missing ⚠️ ](https://app.codecov.io/gh/sillsdev/machine.py/pull/131?src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=sillsdev)
> [...nslation/thot/thot_word_alignment_model_trainer.py](https://app.codecov.io/gh/sillsdev/machine.py/pull/131?src=pr&el=tree&filepath=machine%2Ftranslation%2Fthot%2Fthot_word_alignment_model_trainer.py&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=sillsdev#diff-bWFjaGluZS90cmFuc2xhdGlvbi90aG90L3Rob3Rfd29yZF9hbGlnbm1lbnRfbW9kZWxfdHJhaW5lci5weQ==) 	50.00% 	[2 Missing ⚠️ ](https://app.codecov.io/gh/sillsdev/machine.py/pull/131?src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=sillsdev)
> [machine/corpora/parallel_text_corpus.py](https://app.codecov.io/gh/sillsdev/machine.py/pull/131?src=pr&el=tree&filepath=machine%2Fcorpora%2Fparallel_text_corpus.py&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=sillsdev#diff-bWFjaGluZS9jb3Jwb3JhL3BhcmFsbGVsX3RleHRfY29ycHVzLnB5) 	75.00% 	[1 Missing ⚠️ ](https://app.codecov.io/gh/sillsdev/machine.py/pull/131?src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=sillsdev)
> [machine/jobs/nmt_engine_build_job.py](https://app.codecov.io/gh/sillsdev/machine.py/pull/131?src=pr&el=tree&filepath=machine%2Fjobs%2Fnmt_engine_build_job.py&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=sillsdev#diff-bWFjaGluZS9qb2JzL25tdF9lbmdpbmVfYnVpbGRfam9iLnB5) 	0.00% 	[1 Missing ⚠️ ](https://app.codecov.io/gh/sillsdev/machine.py/pull/131?src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=sillsdev)
> [...translation/huggingface/hugging_face_nmt_engine.py](https://app.codecov.io/gh/sillsdev/machine.py/pull/131?src=pr&el=tree&filepath=machine%2Ftranslation%2Fhuggingface%2Fhugging_face_nmt_engine.py&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=sillsdev#diff-bWFjaGluZS90cmFuc2xhdGlvbi9odWdnaW5nZmFjZS9odWdnaW5nX2ZhY2Vfbm10X2VuZ2luZS5weQ==) 	0.00% 	[1 Missing ⚠️ ](https://app.codecov.io/gh/sillsdev/machine.py/pull/131?src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=sillsdev)
> [...ine/translation/thot/simplex_model_weight_tuner.py](https://app.codecov.io/gh/sillsdev/machine.py/pull/131?src=pr&el=tree&filepath=machine%2Ftranslation%2Fthot%2Fsimplex_model_weight_tuner.py&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=sillsdev#diff-bWFjaGluZS90cmFuc2xhdGlvbi90aG90L3NpbXBsZXhfbW9kZWxfd2VpZ2h0X3R1bmVyLnB5) 	75.00% 	[1 Missing ⚠️ ](https://app.codecov.io/gh/sillsdev/machine.py/pull/131?src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=sillsdev)
> Additional details and impacted files
> 
> [☔ View full report in Codecov by Sentry](https://app.codecov.io/gh/sillsdev/machine.py/pull/131?dropdown=coverage&src=pr&el=continue&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=sillsdev). 📢 Have feedback on the report? [Share it here](https://about.codecov.io/codecov-pr-comment-feedback/?utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=sillsdev).

@ddaspit @johnml1135  I don't understand what this is telling me to change, could one of you explain it to me?

TaperChipmunk32 commented:
Transformers was still not upgraded to the latest version. They introduced some changes in [this pr](https://github.com/huggingface/transformers/pull/29834) that removes some deprecated features that we were using, and I have not found what replaced them yet. We can either go ahead and merge this PR with transformers version 4.41.2, or hold off on it while I figure out how to upgrade transformers to the latest version successfully.

ddaspit reviewed (changes requested):
Do these changes mean that we can no longer support transformers<4.42.0?

> Reviewed 14 of 16 files at r1, 7 of 11 files at r2, 21 of 23 files at r3, 6 of 8 files at r4, 2 of 3 files at r5, 1 of 1 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/sillsdev/machine.py/131#-:-O9vNssq1V_gm39x6up9:bb7x6w6)* status: all files reviewed, 12 unresolved discussions (waiting on @johnml1135, @mshannon-sil, and @TaperChipmunk32)

___
*[`.gitignore` line 141 at r6](https://reviewable.io/reviews/sillsdev/machine.py/131#-O9vDz0h9DWdWY-pVmAR:-O9vDz0h9DWdWY-pVmAS:b-e0tnby) ([raw file](https://github.com/sillsdev/machine.py/blob/3bf63a28cd0db756c2bacd97b2a0801e38555237/.gitignore#L141)):*
> ```ignore
> 
> #VS Code
> .vscode
> ```

The `.vscode` directory shouldn't be ignored.
___
*[`machine/corpora/parallel_text_corpus.py` line 472 at r6](https://reviewable.io/reviews/sillsdev/machine.py/131#-O9vJl0zEJSQ5o05jxWO:-O9vJl0zEJSQ5o05jxWP:b-t78zp4) ([raw file](https://github.com/sillsdev/machine.py/blob/3bf63a28cd0db756c2bacd97b2a0801e38555237/machine/corpora/parallel_text_corpus.py#L472)):*
> ```Python
> 
>         def iterable() -> Iterable[Tuple[Union[str, int], dict]]:
>             key = ""
> ```

Why was this change made?
___
*[`machine/corpora/parallel_text_corpus.py` line 621 at r6](https://reviewable.io/reviews/sillsdev/machine.py/131#-O9vLKqL5_wXajLD2SJE:-O9vLKqL5_wXajLD2SJF:biwvyop) ([raw file](https://github.com/sillsdev/machine.py/blob/3bf63a28cd0db756c2bacd97b2a0801e38555237/machine/corpora/parallel_text_corpus.py#L621)):*
> ```Python
>                 return len(self._df)
>             return len(self._df[(self._df[self._source_column] != "") & (self._df[self._target_column] != "")])
>         return len(self._df[self._df[self._source_column].isin(cast(Sequence[str], text_ids))]) & (
> ```

Why is this necessary?
___
*[`machine/jobs/huggingface/hugging_face_nmt_model_factory.py` line 42 at r6](https://reviewable.io/reviews/sillsdev/machine.py/131#-O9vMmK2EDmqg_lgG3-U:-O9vMmK2EDmqg_lgG3-V:bsz3b73) ([raw file](https://github.com/sillsdev/machine.py/blob/3bf63a28cd0db756c2bacd97b2a0801e38555237/machine/jobs/huggingface/hugging_face_nmt_model_factory.py#L42)):*
> ```Python
>             and "clearml" in self._training_args.report_to
>         ):
>             if isinstance(self._training_args.report_to, List):
> ```

We typically check `isinstance` against the `list` class instead of the `List` type.
___
*[`machine/jobs/huggingface/hugging_face_nmt_model_factory.py` line 45 at r6](https://reviewable.io/reviews/sillsdev/machine.py/131#-O9vMF7i3hTnFE3L5LdH:-O9vMF7i3hTnFE3L5LdI:byj2qs0) ([raw file](https://github.com/sillsdev/machine.py/blob/3bf63a28cd0db756c2bacd97b2a0801e38555237/machine/jobs/huggingface/hugging_face_nmt_model_factory.py#L45)):*
> ```Python
>                 self._training_args.report_to.remove("clearml")
>             elif isinstance(self._training_args.report_to, str):
>                 self._training_args.report_to = None
> ```

You should only set this to `None` if the string is equal to `"clearml"`.
___
*[`machine/translation/thot/simplex_model_weight_tuner.py` line 86 at r6](https://reviewable.io/reviews/sillsdev/machine.py/131#-O9w-ItfAgLW2nryrZVz:-O9w-ItfAgLW2nryrZW-:b6lbhrp) ([raw file](https://github.com/sillsdev/machine.py/blob/3bf63a28cd0db756c2bacd97b2a0801e38555237/machine/translation/thot/simplex_model_weight_tuner.py#L86)):*
> ```Python
>             model = load_smt_model(self._word_alignment_model_type, parameters)
>             decoder = load_smt_decoder(model, parameters)
>             if decoder is not None:
> ```

This is caused by an issue in the Thot type stubs. I will fix that and then these errors should go away.
___
*[`machine/translation/thot/thot_smt_model.py` line 317 at r6](https://reviewable.io/reviews/sillsdev/machine.py/131#-O9w-WXc1g97RoswKAZT:-O9w-WXc1g97RoswKAZU:bdhf0e) ([raw file](https://github.com/sillsdev/machine.py/blob/3bf63a28cd0db756c2bacd97b2a0801e38555237/machine/translation/thot/thot_smt_model.py#L317)):*
> ```Python
>                 wa_matrix = WordAlignmentMatrix.from_word_pairs(1, 1, {(0, 0)})
>             else:
>                 src_phrase = list(normalized_source_tokens[source_start_index : source_start_index + src_phrase_len])
> ```

Why was this necessary?
___
*[`tests/corpora/test_dbl_bundle_text_corpus.py` line 1 at r6](https://reviewable.io/reviews/sillsdev/machine.py/131#-O9vztVx-OKoFpaKJxkZ:-O9vztVy6SJjviJ99tmg:byyw5ww) ([raw file](https://github.com/sillsdev/machine.py/blob/3bf63a28cd0db756c2bacd97b2a0801e38555237/tests/corpora/test_dbl_bundle_text_corpus.py#L1)):*
> ```Python
> from tests.testutils.dbl_bundle_test_environment import DblBundleTestEnvironment
> ```

If you add `extraPaths = ["tests"]` in the `tools.pyright` section of the pyproject.toml file, then it will fix all of these errors.


<!-- Sent from Reviewable.io -->

johnml1135 commented:
*[`.devcontainer/devcontainer.json` line 12 at r6](https://reviewable.io/reviews/sillsdev/machine.py/131#-O9yYST34Nk7uJ1wIt0D:-O9yYST34Nk7uJ1wIt0E:br8trw0) ([raw file](https://github.com/sillsdev/machine.py/blob/3bf63a28cd0db756c2bacd97b2a0801e38555237/.devcontainer/devcontainer.json#L12)):*
> ```jsonc
> 	},
> 	"runArgs": [
> 		//"--gpus",
> ```

This should be reversed.  If you want, create a CPU only dev environment that doesn't have the runArgs.


<!-- Sent from Reviewable.io -->

johnml1135 commented:
*[`machine/corpora/parallel_text_corpus.py` line 621 at r6](https://reviewable.io/reviews/sillsdev/machine.py/131#-O9vLKqL5_wXajLD2SJE:-O9yZ-YN3R_JAQKecOC-:bn76v4j) ([raw file](https://github.com/sillsdev/machine.py/blob/3bf63a28cd0db756c2bacd97b2a0801e38555237/machine/corpora/parallel_text_corpus.py#L621)):*
<details><summary><i>Previously, ddaspit (Damien Daspit) wrote…</i></summary><blockquote>

Why is this necessary?
</blockquote></details>

We're these changes due to the updates in python?


<!-- Sent from Reviewable.io -->

johnml1135 commented:
*[`machine/translation/huggingface/hugging_face_nmt_model_trainer.py` line 230 at r6](https://reviewable.io/reviews/sillsdev/machine.py/131#-O9yZNPH-uaY7f8kyh-J:-O9yZNPH-uaY7f8kyh-K:b-xiweex) ([raw file](https://github.com/sillsdev/machine.py/blob/3bf63a28cd0db756c2bacd97b2a0801e38555237/machine/translation/huggingface/hugging_face_nmt_model_trainer.py#L230)):*
> ```Python
>             tokenizer.lang_code_to_id[lang_code] = lang_id
> 
>             if isinstance(tokenizer, (MBart50Tokenizer, MBartTokenizer)):
> ```

Why are these changes needed?


<!-- Sent from Reviewable.io -->

johnml1135 reviewed (commented):
> Reviewed 9 of 16 files at r1, 5 of 11 files at r2, 3 of 8 files at r4, 2 of 3 files at r5, 1 of 1 files at r6, 30 of 30 files at r7, 1 of 1 files at r8, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/sillsdev/machine.py/131)* status: all files reviewed, 13 unresolved discussions (waiting on @ddaspit, @mshannon-sil, and @TaperChipmunk32)



<!-- Sent from Reviewable.io -->

TaperChipmunk32 commented:
@ddaspit @johnml1135 Huggingface just released transformers version 4.46.0, which deprecates the argument "tokenizer" and renames it to "processing_class" for Trainer. This breaks some of our code that uses the Seq2SeqTrainer. So for now I am limiting support to >=4.38.0 and <4.46.0.

ddaspit reviewed (commented):
Let's update it to support `transformers = ">=4.38.0,<4.46.0"`. Also, you updated the sentencepiece, datasets, and sacremoses versions. Was that necessary?

> Reviewed 28 of 30 files at r7, 1 of 1 files at r8, 2 of 2 files at r9, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/sillsdev/machine.py/131#-:-OA-E9MeEoYSnFWAglLi:bg3rl8m)* status: all files reviewed, 10 unresolved discussions (waiting on @johnml1135, @mshannon-sil, and @TaperChipmunk32)

___
*[`pyproject.toml` line 67 at r1](https://reviewable.io/reviews/sillsdev/machine.py/131#-O9WPe_oAqte-xkCwsaI:-O9yrnTRB-VmYaZSCW12:b-v3ep23) ([raw file](https://github.com/sillsdev/machine.py/blob/1c47de3b2062f500535f4acd2448915a9a36a6f8/pyproject.toml#L67)):*
<details><summary><i>Previously, TaperChipmunk32 (Matthew Beech) wrote…</i></summary><blockquote>

I was going to keep transformers downgraded because of some changes made to the NLLBTokenizers in later versions, but I will work on upgrading it instead. 

I've also started looking into upgrading the other packages mentioned.

</blockquote></details>

There is no need to upgrade the packages if they work. The versions specified here are intended to be the minimum version that Machine supports.
___
*[`pyproject.toml` line 66 at r9](https://reviewable.io/reviews/sillsdev/machine.py/131#-OA-ErFTB9pgw5fnjcA1:-OA-ErFTB9pgw5fnjcA2:bz1yl0q) ([raw file](https://github.com/sillsdev/machine.py/blob/81c52e7d1b766fc77e8752daec2f6b9306b314c7/pyproject.toml#L66)):*
> ```TOML
> sil-thot = "^3.4.4"
> 
> transformers = "^4.38.0, <4.46.0"
> ```

The version should be specified as `>=4.38.0,<4.46.0`.
___
*[`machine/corpora/parallel_text_corpus.py` line 621 at r6](https://reviewable.io/reviews/sillsdev/machine.py/131#-O9vLKqL5_wXajLD2SJE:-OA-DNQ73C4q2UY02WWF:b-ve4dmk) ([raw file](https://github.com/sillsdev/machine.py/blob/3bf63a28cd0db756c2bacd97b2a0801e38555237/machine/corpora/parallel_text_corpus.py#L621)):*
<details><summary><i>Previously, johnml1135 (John Lambert) wrote…</i></summary><blockquote>

We're these changes due to the updates in python?
</blockquote></details>

I don't receive a type error here. What error are you getting?
___
*[`machine/translation/thot/simplex_model_weight_tuner.py` line 86 at r6](https://reviewable.io/reviews/sillsdev/machine.py/131#-O9w-ItfAgLW2nryrZVz:-OA-66Yp1aYsCM4dX2-0:b89tzyq) ([raw file](https://github.com/sillsdev/machine.py/blob/3bf63a28cd0db756c2bacd97b2a0801e38555237/machine/translation/thot/simplex_model_weight_tuner.py#L86)):*
<details><summary><i>Previously, ddaspit (Damien Daspit) wrote…</i></summary><blockquote>

This is caused by an issue in the Thot type stubs. I will fix that and then these errors should go away.
</blockquote></details>

There is a new version of Thot (3.4.5) that should fix these issues. Also, you should add `reportMissingModuleSource = false` in the `tool.pyright` section of the pyproject.toml. It will get rid of the spurious warnings when importing Thot modules.
___
*[`machine/translation/thot/thot_smt_model.py` line 319 at r9](https://reviewable.io/reviews/sillsdev/machine.py/131#-OA-DCOBB3i0hBxITsQF:-OA-DCOBB3i0hBxITsQG:b-ve4dmk) ([raw file](https://github.com/sillsdev/machine.py/blob/81c52e7d1b766fc77e8752daec2f6b9306b314c7/machine/translation/thot/thot_smt_model.py#L319)):*
> ```Python
>                 src_phrase = [""] * src_phrase_len
>                 for i in range(src_phrase_len):
>                     src_phrase[i] = normalized_source_tokens[source_start_index + i]  # type: ignore
> ```

I don't receive a type error here. What error are you getting?
___
*[`tests/corpora/test_dbl_bundle_text_corpus.py` line 1 at r6](https://reviewable.io/reviews/sillsdev/machine.py/131#-O9vztVx-OKoFpaKJxkZ:-OA-4GEZBbnlzTjq4zMp:b-c8q6zh) ([raw file](https://github.com/sillsdev/machine.py/blob/3bf63a28cd0db756c2bacd97b2a0801e38555237/tests/corpora/test_dbl_bundle_text_corpus.py#L1)):*
<details><summary><i>Previously, TaperChipmunk32 (Matthew Beech) wrote…</i></summary><blockquote>

1. I'll remove .vscode from .gitignore
2. It was left over from another change I made to try and remove an error. It was throwing another error about all code paths must return the correct type and not None. I was able to revert it with no issue.
3. isin does not accept set[str], it only accepts Series | DataFrame | Sequence[Unknown] | Mapping[Unknown, Unknown]
4. I'll fix that
5. I'll fix that
6. Alright, I'll wait on those changes to Thot then and revert my changes
7. An error was occurring about "Argument of type "Sequence[str]" cannot be assigned to parameter "value" of type "str"". Another option is to just add type: ignore and tests still pass.
8. I'll add that in

</blockquote></details>

Can you reply to each individual comment? It is difficult for me to figure out what you are replying to for each comment here.


<!-- Sent from Reviewable.io -->

TaperChipmunk32 commented:
> Do these changes mean that we can no longer support transformers<4.42.0?
> 
> > Reviewed 14 of 16 files at r1, 7 of 11 files at r2, 21 of 23 files at r3, 6 of 8 files at r4, 2 of 3 files at r5, 1 of 1 files at r6, all commit messages.
> > _[Reviewable](https://reviewable.io/reviews/sillsdev/machine.py/131#-:-O9vNssq1V_gm39x6up9:bb7x6w6)_ status: all files reviewed, 12 unresolved discussions (waiting on @johnml1135, @mshannon-sil, and @TaperChipmunk32)
> 
> _[`.gitignore` line 141 at r6](https://reviewable.io/reviews/sillsdev/machine.py/131#-O9vDz0h9DWdWY-pVmAR:-O9vDz0h9DWdWY-pVmAS:b-e0tnby) ([raw file](https://github.com/sillsdev/machine.py/blob/3bf63a28cd0db756c2bacd97b2a0801e38555237/.gitignore#L141)):_
> 
> > ```gitignore
> > #VS Code
> > .vscode
> > ```
> 
> The `.vscode` directory shouldn't be ignored.

I've removed it now

> 
> _[`machine/corpora/parallel_text_corpus.py` line 472 at r6](https://reviewable.io/reviews/sillsdev/machine.py/131#-O9vJl0zEJSQ5o05jxWO:-O9vJl0zEJSQ5o05jxWP:b-t78zp4) ([raw file](https://github.com/sillsdev/machine.py/blob/3bf63a28cd0db756c2bacd97b2a0801e38555237/machine/corpora/parallel_text_corpus.py#L472)):_
> 
> > ```python
> >         def iterable() -> Iterable[Tuple[Union[str, int], dict]]:
> >             key = ""
> > ```
> 
> Why was this change made?

It was left over from another change I made to try and remove an error. It was throwing another error about all code paths must return the correct type and not None. I was able to revert it with no issue.

> _[`machine/corpora/parallel_text_corpus.py` line 621 at r6](https://reviewable.io/reviews/sillsdev/machine.py/131#-O9vLKqL5_wXajLD2SJE:-O9vLKqL5_wXajLD2SJF:biwvyop) ([raw file](https://github.com/sillsdev/machine.py/blob/3bf63a28cd0db756c2bacd97b2a0801e38555237/machine/corpora/parallel_text_corpus.py#L621)):_
> 
> > ```python
> >                 return len(self._df)
> >             return len(self._df[(self._df[self._source_column] != "") & (self._df[self._target_column] != "")])
> >         return len(self._df[self._df[self._source_column].isin(cast(Sequence[str], text_ids))]) & (
> > ```
> 
> Why is this necessary?

isin does not accept set[str], it only accepts Series | DataFrame | Sequence[Unknown] | Mapping[Unknown, Unknown]

> _[`machine/jobs/huggingface/hugging_face_nmt_model_factory.py` line 42 at r6](https://reviewable.io/reviews/sillsdev/machine.py/131#-O9vMmK2EDmqg_lgG3-U:-O9vMmK2EDmqg_lgG3-V:bsz3b73) ([raw file](https://github.com/sillsdev/machine.py/blob/3bf63a28cd0db756c2bacd97b2a0801e38555237/machine/jobs/huggingface/hugging_face_nmt_model_factory.py#L42)):_
> 
> > ```python
> >             and "clearml" in self._training_args.report_to
> >         ):
> >             if isinstance(self._training_args.report_to, List):
> > ```
> 
> We typically check `isinstance` against the `list` class instead of the `List` type.

I fixed this

> _[`machine/jobs/huggingface/hugging_face_nmt_model_factory.py` line 45 at r6](https://reviewable.io/reviews/sillsdev/machine.py/131#-O9vMF7i3hTnFE3L5LdH:-O9vMF7i3hTnFE3L5LdI:byj2qs0) ([raw file](https://github.com/sillsdev/machine.py/blob/3bf63a28cd0db756c2bacd97b2a0801e38555237/machine/jobs/huggingface/hugging_face_nmt_model_factory.py#L45)):_
> 
> > ```python
> >                 self._training_args.report_to.remove("clearml")
> >             elif isinstance(self._training_args.report_to, str):
> >                 self._training_args.report_to = None
> > ```
> 
> You should only set this to `None` if the string is equal to `"clearml"`.

I fixed this

> _[`machine/translation/thot/simplex_model_weight_tuner.py` line 86 at r6](https://reviewable.io/reviews/sillsdev/machine.py/131#-O9w-ItfAgLW2nryrZVz:-O9w-ItfAgLW2nryrZW-:b6lbhrp) ([raw file](https://github.com/sillsdev/machine.py/blob/3bf63a28cd0db756c2bacd97b2a0801e38555237/machine/translation/thot/simplex_model_weight_tuner.py#L86)):_
> 
> > ```python
> >             model = load_smt_model(self._word_alignment_model_type, parameters)
> >             decoder = load_smt_decoder(model, parameters)
> >             if decoder is not None:
> > ```
> 
> This is caused by an issue in the Thot type stubs. I will fix that and then these errors should go away.

Alright, I updated Thot to the newest version

> _[`machine/translation/thot/thot_smt_model.py` line 317 at r6](https://reviewable.io/reviews/sillsdev/machine.py/131#-O9w-WXc1g97RoswKAZT:-O9w-WXc1g97RoswKAZU:bdhf0e) ([raw file](https://github.com/sillsdev/machine.py/blob/3bf63a28cd0db756c2bacd97b2a0801e38555237/machine/translation/thot/thot_smt_model.py#L317)):_
> 
> > ```python
> >                 wa_matrix = WordAlignmentMatrix.from_word_pairs(1, 1, {(0, 0)})
> >             else:
> >                 src_phrase = list(normalized_source_tokens[source_start_index : source_start_index + src_phrase_len])
> > ```
> 
> Why was this necessary?

An error was occurring about "Argument of type "Sequence[str]" cannot be assigned to parameter "value" of type "str"". Another option is to just add type: ignore and tests still pass.


> _[`tests/corpora/test_dbl_bundle_text_corpus.py` line 1 at r6](https://reviewable.io/reviews/sillsdev/machine.py/131#-O9vztVx-OKoFpaKJxkZ:-O9vztVy6SJjviJ99tmg:byyw5ww) ([raw file](https://github.com/sillsdev/machine.py/blob/3bf63a28cd0db756c2bacd97b2a0801e38555237/tests/corpora/test_dbl_bundle_text_corpus.py#L1)):_
> 
> > ```python
> > from tests.testutils.dbl_bundle_test_environment import DblBundleTestEnvironment
> > ```
> 
> If you add `extraPaths = ["tests"]` in the `tools.pyright` section of the pyproject.toml file, then it will fix all of these errors.

I've added that and reverted those changes.

TaperChipmunk32 commented:
> Let's update it to support `transformers = ">=4.38.0,<4.46.0"`. Also, you updated the sentencepiece, datasets, and sacremoses versions. Was that necessary?

I've updated support of transformers to those versions. The other updates were not necessary, so I reverted the change.

> 
> > Reviewed 28 of 30 files at r7, 1 of 1 files at r8, 2 of 2 files at r9, all commit messages.
> > _[Reviewable](https://reviewable.io/reviews/sillsdev/machine.py/131#-:-OA-E9MeEoYSnFWAglLi:bg3rl8m)_ status: all files reviewed, 10 unresolved discussions (waiting on @johnml1135, @mshannon-sil, and @TaperChipmunk32)
> 
> _[`pyproject.toml` line 67 at r1](https://reviewable.io/reviews/sillsdev/machine.py/131#-O9WPe_oAqte-xkCwsaI:-O9yrnTRB-VmYaZSCW12:b-v3ep23) ([raw file](https://github.com/sillsdev/machine.py/blob/1c47de3b2062f500535f4acd2448915a9a36a6f8/pyproject.toml#L67)):_
> Previously, TaperChipmunk32 (Matthew Beech) wrote…
> 
> There is no need to upgrade the packages if they work. The versions specified here are intended to be the minimum version that Machine supports.
>
> _[`pyproject.toml` line 66 at r9](https://reviewable.io/reviews/sillsdev/machine.py/131#-OA-ErFTB9pgw5fnjcA1:-OA-ErFTB9pgw5fnjcA2:bz1yl0q) ([raw file](https://github.com/sillsdev/machine.py/blob/81c52e7d1b766fc77e8752daec2f6b9306b314c7/pyproject.toml#L66)):_
> 
> > ```toml
> > sil-thot = "^3.4.4"
> > 
> > transformers = "^4.38.0, <4.46.0"
> > ```
> 
> The version should be specified as `>=4.38.0,<4.46.0`.
> 
> _[`machine/corpora/parallel_text_corpus.py` line 621 at r6](https://reviewable.io/reviews/sillsdev/machine.py/131#-O9vLKqL5_wXajLD2SJE:-OA-DNQ73C4q2UY02WWF:b-ve4dmk) ([raw file](https://github.com/sillsdev/machine.py/blob/3bf63a28cd0db756c2bacd97b2a0801e38555237/machine/corpora/parallel_text_corpus.py#L621)):_
> Previously, johnml1135 (John Lambert) wrote…
> 
> I don't receive a type error here. What error are you getting?

isin does not accept set[str], it only accepts Series | DataFrame | Sequence[Unknown] | Mapping[Unknown, Unknown]. I do not get the error on vscode, but in pyright during the build process.

> 
> _[`machine/translation/thot/simplex_model_weight_tuner.py` line 86 at r6](https://reviewable.io/reviews/sillsdev/machine.py/131#-O9w-ItfAgLW2nryrZVz:-OA-66Yp1aYsCM4dX2-0:b89tzyq) ([raw file](https://github.com/sillsdev/machine.py/blob/3bf63a28cd0db756c2bacd97b2a0801e38555237/machine/translation/thot/simplex_model_weight_tuner.py#L86)):_
> Previously, ddaspit (Damien Daspit) wrote…
> 
> There is a new version of Thot (3.4.5) that should fix these issues. Also, you should add `reportMissingModuleSource = false` in the `tool.pyright` section of the pyproject.toml. It will get rid of the spurious warnings when importing Thot modules.

Done, I've updated Thot and added that line to pyproject.toml. I also reverted the changes I made caused by the Thot typing issues.

> _[`machine/translation/thot/thot_smt_model.py` line 319 at r9](https://reviewable.io/reviews/sillsdev/machine.py/131#-OA-DCOBB3i0hBxITsQF:-OA-DCOBB3i0hBxITsQG:b-ve4dmk) ([raw file](https://github.com/sillsdev/machine.py/blob/81c52e7d1b766fc77e8752daec2f6b9306b314c7/machine/translation/thot/thot_smt_model.py#L319)):_
> 
> > ```python
> >                 src_phrase = [""] * src_phrase_len
> >                 for i in range(src_phrase_len):
> >                     src_phrase[i] = normalized_source_tokens[source_start_index + i]  # type: ignore
> > ```
> 
> I don't receive a type error here. What error are you getting?

I also can no longer recreate the error, I'm not sure what was causing it.

> 
> _[`tests/corpora/test_dbl_bundle_text_corpus.py` line 1 at r6](https://reviewable.io/reviews/sillsdev/machine.py/131#-O9vztVx-OKoFpaKJxkZ:-OA-4GEZBbnlzTjq4zMp:b-c8q6zh) ([raw file](https://github.com/sillsdev/machine.py/blob/3bf63a28cd0db756c2bacd97b2a0801e38555237/tests/corpora/test_dbl_bundle_text_corpus.py#L1)):_
> Previously, TaperChipmunk32 (Matthew Beech) wrote…
> 
> Can you reply to each individual comment? It is difficult for me to figure out what you are replying to for each comment here.

I've deleted that old reply and made another that is more readable.

ddaspit reviewed (commented):
> Reviewed 5 of 5 files at r10, 1 of 2 files at r11, 1 of 1 files at r12, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/sillsdev/machine.py/131)* status: all files reviewed, 6 unresolved discussions (waiting on @johnml1135, @mshannon-sil, and @TaperChipmunk32)

___
*[`machine/corpora/parallel_text_corpus.py` line 621 at r6](https://reviewable.io/reviews/sillsdev/machine.py/131#-O9vLKqL5_wXajLD2SJE:-OAIp_rf59sCcJblQjuh:bxp5vbz) ([raw file](https://github.com/sillsdev/machine.py/blob/3bf63a28cd0db756c2bacd97b2a0801e38555237/machine/corpora/parallel_text_corpus.py#L621)):*
<details><summary><i>Previously, TaperChipmunk32 (Matthew Beech) wrote…</i></summary><blockquote>

isin does not accept set[str], it only accepts Series | DataFrame | Sequence[Unknown] | Mapping[Unknown, Unknown]. I do not get the error on vscode, but in pyright during the build process.
</blockquote></details>

Ok, I was able to reproduce the issue. Rather than casting to `Sequence`, you should convert `text_ids` to a list and then pass it in to the `isin` calls.


<!-- Sent from Reviewable.io -->

ddaspit reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 1 of 1 files at r13, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/sillsdev/machine.py/131#-:-OAIuRn_1FC00BhV851m:bnfp4nl)* status: all files reviewed, 5 unresolved discussions (waiting on @johnml1135, @mshannon-sil, and @TaperChipmunk32)



<!-- Sent from Reviewable.io -->

TaperChipmunk32 commented:
@ddaspit @johnml1135 I keep getting an error with Windows directory permission denied that I cannot seem to fix, does anyone have any ideas? You can see the error in the recent build failure. All other OS builds complete successfully.

ddaspit reviewed (commented):
I figured out the issue. It looks like the trainer and engine aren't getting cleaned up properly in the unit test. I will push a commit with the fix.

> Reviewed all commit messages.
> *[Reviewable](https://reviewable.io/reviews/sillsdev/machine.py/131#-:-OAJTHFy9ST_ewaOhRqX:b-ktvv8d)* status: all files reviewed, 5 unresolved discussions (waiting on @johnml1135, @mshannon-sil, and @TaperChipmunk32)



<!-- Sent from Reviewable.io -->

johnml1135 commented:
*[`pyproject.toml` line 67 at r1](https://reviewable.io/reviews/sillsdev/machine.py/131#-O9WPe_oAqte-xkCwsaI:-OAJXAgSD86orpkOff6F:b2ni9bh) ([raw file](https://github.com/sillsdev/machine.py/blob/1c47de3b2062f500535f4acd2448915a9a36a6f8/pyproject.toml#L67)):*
<details><summary><i>Previously, ddaspit (Damien Daspit) wrote…</i></summary><blockquote>

There is no need to upgrade the packages if they work. The versions specified here are intended to be the minimum version that Machine supports.
</blockquote></details>

ok.


<!-- Sent from Reviewable.io -->

johnml1135 reviewed (commented):
> Reviewed 1 of 5 files at r10.
> *[Reviewable](https://reviewable.io/reviews/sillsdev/machine.py/131)* status: 44 of 48 files reviewed, 5 unresolved discussions (waiting on @ddaspit, @mshannon-sil, and @TaperChipmunk32)



<!-- Sent from Reviewable.io -->

ddaspit reviewed (commented):
Looks like you caught the unit test that I missed. Also, can you update to the latest version of the actions in the CI build to get rid of the warnings.

> Reviewed 4 of 4 files at r15, 6 of 6 files at r16, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/sillsdev/machine.py/131#-:-OAJWJtU0YWWVLbDnRJ0:b8eomqz)* status: all files reviewed, 5 unresolved discussions (waiting on @mshannon-sil and @TaperChipmunk32)



<!-- Sent from Reviewable.io -->

TaperChipmunk32 commented:
> Looks like you caught the unit test that I missed. Also, can you update to the latest version of the actions in the CI build to get rid of the warnings.
> 
> > Reviewed 4 of 4 files at r15, 6 of 6 files at r16, all commit messages.
> > _[Reviewable](https://reviewable.io/reviews/sillsdev/machine.py/131#-:-OAJWJtU0YWWVLbDnRJ0:b8eomqz)_ status: all files reviewed, 5 unresolved discussions (waiting on @mshannon-sil and @TaperChipmunk32)

I updated github actions and the warnings are now gone.

ddaspit reviewed (commented):
> Reviewed 3 of 5 files at r17, 32 of 32 files at r18, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/sillsdev/machine.py/131)* status: all files reviewed, 5 unresolved discussions (waiting on @mshannon-sil and @TaperChipmunk32)



<!-- Sent from Reviewable.io -->

johnml1135 commented:
@mshannon-sil - are you ok with all the changes?





<!-- Sent from Reviewable.io -->

mshannon-sil reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 2 of 11 files at r2, 1 of 30 files at r7, 1 of 5 files at r10, 1 of 1 files at r13, 2 of 4 files at r15, 1 of 6 files at r16, 3 of 5 files at r17, 32 of 32 files at r18, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/sillsdev/machine.py/131#-:-OAOXdU3-GrbRGNAcYsw:bnfp4nl)* status: all files reviewed, 2 unresolved discussions (waiting on @TaperChipmunk32)



<!-- Sent from Reviewable.io -->

TaperChipmunk32 commented:
This PR is ready to be merged. When this is merged, do I need to send out a message telling everyone to rebuild their dev container and update their poetry environment? @ddaspit
